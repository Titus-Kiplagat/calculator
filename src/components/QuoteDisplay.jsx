import { useEffect, useState } from 'react';

function QuoteDisplay() {
  const [quote, setQuote] = useState('');
  const [author, setAuthor] = useState('');
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const timeout = setTimeout(() => {
      const fetchQuote = async () => {
        const quoteUrl = process.env.REACT_APP_QUOTE_URL;
        const apiKey = process.env.REACT_APP_API_KEY;
        const category = 'learning';

        try {
          const response = await fetch(`${quoteUrl}?category=${category}`, { headers: { 'X-Api-Key': `${apiKey}` } });
          const data = await response.json();

          if (!response.ok) {
            throw new Error('Failed to fetch quote.');
          }

          const { quote, author } = data[0];
          setQuote(quote);
          setAuthor(author);
        } catch (error) {
          setError(error.message);
        } finally {
          setLoading(false);
        }
      };

      fetchQuote();
    }, 1000);

    return () => clearTimeout(timeout);
  }, []);

  if (loading) {
    return <div className="mt-10">Loading...</div>;
  }

  if (error) {
    return (
      <div>
        Error:
        {' '}
        {error}
      </div>
    );
  }

  return (
    <div className="bg-gray-100 w-2/4 p-5 m-5 text-center rounded shadow-md shadow-gray-500 text-orange-900 text-lg font-black">
      <h3>
        &#34;
        {quote}
        &#34;
        {' '}
        by
        {' '}
        <span>
          &#34;
          {author}
          &#34;
        </span>
      </h3>
    </div>
  );
}

export default QuoteDisplay;
