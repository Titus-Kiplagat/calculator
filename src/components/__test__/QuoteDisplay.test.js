import { render } from '@testing-library/react';
import QuoteDisplay from '../QuoteDisplay';

global.fetch = jest.fn(() => Promise.resolve({
  ok: true,
  json: () => Promise.resolve([{ quote: 'Test quote', author: 'Test author' }]),
}));

describe('QuoteDisplay', () => {
  it('QuoteDisplay component matches snapshot', () => {
    const { asFragment } = render(<QuoteDisplay />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render loading message when fetching quote', () => {
    // Mock the fetch function
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      ok: true,
      json: () => Promise.resolve([{ quote: 'Test quote', author: 'Test author' }]),
    }));

    // Render the component
    const { getByText } = render(<QuoteDisplay />);

    // Assert that the loading message is rendered
    expect(getByText('Loading...')).toBeInTheDocument();

    // Clean up the mock
    global.fetch.mockRestore();
  });
});
