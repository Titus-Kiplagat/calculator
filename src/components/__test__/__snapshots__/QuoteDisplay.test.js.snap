// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteDisplay QuoteDisplay component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-5/6 flex justify-center items-center"
  >
    <div
      class="mt-10"
    >
      Loading...
    </div>
  </div>
</DocumentFragment>
`;
